src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -1701,16 +1707,47 @@ private IEnumerable<PSTypeName> InferTypesFrom(MemberExpressionAst memberExpress
 
         private static void AddTypesFromMethodCacheEntry(
             DotNetAdapter.MethodCacheEntry methodCacheEntry,
+            IList<ITypeName> genericTypeArguments,
             List<PSTypeName> result,
             bool isInvokeMemberExpressionAst)
         {
             if (isInvokeMemberExpressionAst)
             {
+                Type[] resolvedTypeArguments = null;
+                if (genericTypeArguments is not null)
+                {
+                    resolvedTypeArguments = new Type[genericTypeArguments.Count];
+                    for (int i = 0; i < genericTypeArguments.Count; i++)
+                    {
+                        Type resolvedType = genericTypeArguments[i].GetReflectionType();
+                        if (resolvedType is null)
+                        {
+                            return;
+                        }
+
+                        resolvedTypeArguments[i] = resolvedType;
+                    }
+                }
+
                 foreach (var method in methodCacheEntry.methodInformationStructures)
                 {
-                    if (method.method is MethodInfo methodInfo && !methodInfo.ReturnType.ContainsGenericParameters)
+                    if (method.method is MethodInfo methodInfo)
                     {
-                        result.Add(new PSTypeName(methodInfo.ReturnType));
+                        if (!methodInfo.ReturnType.ContainsGenericParameters)
+                        {
+                            result.Add(new PSTypeName(methodInfo.ReturnType));
+                        }
+                        else if (genericTypeArguments is not null)
+                        {
+                            try
+                            {
+                                result.Add(new PSTypeName(methodInfo.MakeGenericMethod(resolvedTypeArguments).ReturnType));
+                            }
+                            catch // If we can't build the generic method then just skip it to retain other completion results.
+                            {
+                                return;
```

Comment:
Would it be worth logging the failure?

Reply:
I don't think so. I'd be surprised if we log any tab completion / type inference failures.